### PR TITLE
Check transaction signatures in entry verify

### LINKED ISSUE #8593: TVU does not verify transaction signatures
The leader can fabricate transactions with invalid signatures and transmit them to the other validator nodes for verification.  Because those other validator nodes do not perform a transaction signature check (I swear we did this at some point in the past), whenever your node is leader you can issue any transactions you want.

### PR DESCRIPTION
**Pull request by sakridge**
#### Problem

Transaction signatures not verified.

#### Summary of Changes

Verify them in entry verification.

Fixes #8593

### CODE REVIEW

**Comment by garious** [comment]
That's a PoH Verify function. I would have expected it added in the TVU somewhere OR see SigVerifyStage moved to some shared location. Also, surprised to see CPU-based signature verification here.  But given the central location, does this indicate that `ledger-tool` was skipping the verification as well?

I think it'd be best to address the architectural diagrams first on this one:

https://github.com/solana-labs/solana/blob/master/docs/art/validator.bob
https://github.com/solana-labs/solana/blob/master/docs/art/tpu.bob
https://github.com/solana-labs/solana/blob/master/docs/art/tvu.bob

**Comment by mvines** [comment]
> does this indicate that ledger-tool was skipping the verification as well?

Yes, `ledger-tool verify` was not verifying the transaction signatures themselves.  Arguably because by transactions should be all good by the time they hit blockstore, but yeah

**Comment by sakridge** [comment]
> That's a PoH Verify function.

It was, now it's poh verify & transaction signature verify. There's nothing that says poh_verify that I noticed.

> I would have expected it added in the TVU somewhere OR see SigVerifyStage moved to some
> shared location. Also, surprised to see CPU-based signature verification here. 

We can do it in a follow-up change, but this is the easiest/most straightforward way to add it. I did some perf sanity checks, we are still above 50k tps cpu-only on colo.

> But given the central location, does this indicate that `ledger-tool` was skipping the verification as well?

Yes. This also solves this problem.

**Comment by garious** [comment]
This is an architecture bug more than it is an implementation bug. I don't think we should merge this until a change to architecture documents is approved.  Zooming way, way in, all the way to function level (which is already too far), this patch makes the architectural decision to do signature verification before PoH verification. That alone looks like an attack vector (malicious leader randomly resembles transactions and floods the validators).  At the very least, the code in this patch should be in a separate function, so that it's clear one form of verification is happening before the other.

File: ledger/src/entry.rs [hunk 1]
```diff
@@ -509,6 +529,36 @@ mod tests {
...
+
+        // First, verify entries
+        let keypair = Keypair::new();
+        let tx0 = system_transaction::transfer(&keypair, &keypair.pubkey(), 0, zero);
+        let tx1 = system_transaction::transfer(&keypair, &keypair.pubkey(), 1, zero);
+        tx0.verify().unwrap();
+        let mut e0 = vec![Entry::new(&zero, 0, vec![tx0.clone(), tx1.clone()])];
+        assert!(e0.verify(&zero));
+
+        let orig_sig = e0[0].transactions[0].signatures[0];
```
Comment (garious) [inline ledger/src/entry.rs]: comment?

File: ledger/src/entry.rs [hunk 2]
```diff
@@ -509,6 +529,36 @@ mod tests {
...
+        e0[0].transactions[0].signatures[0] = Signature::default();
+        assert!(!e0.verify(&zero));
+
+        e0[0].transactions[0].signatures[0] = orig_sig;
+        assert!(e0.verify(&zero));
+
+        let len = e0[0].transactions[0].signatures.len();
```
Comment (garious) [inline ledger/src/entry.rs]: comment?

**Comment by sakridge** [comment]
> That alone looks like an attack vector (malicious leader randomly resembles transactions and floods the validators)

Why couldn't/wouldn't the malicious leader assemble those transaction to just pass poh verify? It just needs to do some hashing.

**Comment by garious** [comment]
The intent of my comment wasn't to debate an attack vector. It was to give a concrete example of how this innocent looking PR may have significant implications, and how a PR to an architectural document would be a better venue for that discussion.

**Pull request merged**